marshal.go

```diff
@@ -295,15 +306,39 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
 				continue
 			}
 
-			// Detect usmStats report PDUs and go out of this function with all data
-			// (usmStatsNotInTimeWindows [1.3.6.1.6.3.15.1.1.2.0] will be handled by the calling
-			// function, and retransmitted.  All others need to be handled by user code)
-			if result.Version == Version3 && len(result.Variables) == 1 && result.PDUType == Report {
+			// While Report PDU was defined by RFC 1905 as part of SNMPv2, it was never
+			// used until SNMPv3. Report PDU's allow a SNMP engine to tell another SNMP
+			// engine that an error was detected while processing an SNMP message.
+			//
+			// The format for a Report PDU is
+			// -----------------------------------
+			// | 0xA8 | reqid | 0 | 0 | varbinds |
+			// -----------------------------------
+			// where:
+			// - PDU type 0xA8 indicates a Report PDU.
+			// - reqid is either:
+			//    The request identifier of the message that triggered the report
+			//    or zero if the request identifier cannot be extracted.
+			// - The variable bindings will contain a single object identifier and its value
+			//
+			// usmStatsNotInTimeWindows and usmStatsUnknownEngineIDs are recoverable errors
+			// and will be retransmitted, for others we return the result with an error.
+			if result.Version == Version3 && result.PDUType == Report && len(result.Variables) == 1 {
 				switch result.Variables[0].Name {
-				case ".1.3.6.1.6.3.15.1.1.1.0", ".1.3.6.1.6.3.15.1.1.2.0",
-					".1.3.6.1.6.3.15.1.1.3.0", ".1.3.6.1.6.3.15.1.1.4.0",
-					".1.3.6.1.6.3.15.1.1.5.0", ".1.3.6.1.6.3.15.1.1.6.0":
+				case usmStatsUnsupportedSecLevels:
+					return result, fmt.Errorf("error: unknown security level")
+				case usmStatsNotInTimeWindows:
 					break waitingResponse
+				case usmStatsUnknownUserNames:
+					return result, fmt.Errorf("error: unknown username")
```

Comment:
I wonder if we should make these error messages package level exported vars, so they can be used with `errors.Is()`.

Something like
```go
var (
	ErrUnknownUsername = errors.New("unknown username")
...
```

Also while we're here, let's drop the `error: ` text from these messages.

Reply:
This is a nice idea :) Added as you suggested

Reply:
I like this as well. Does that mean that other tools can do this?
```go
result, err := gs.Get([]string{oid})
if err == gosnmp.ErrUnknownSecurityLevel {
    // do something related
}
```

Reply:
> Does that mean that other tools can do this?
> ``if err == gosnmp.ErrUnknownSecurityLevel {``

That should be the case indeed, reading up on the [docs](https://golang.org/pkg/errors/) it's mentioned to use
```
if errors.Is(err, gosnmp.ErrUnknownSecurityLevel)
```
preferable to
```
if err == gosnmp.ErrUnknownSecurityLevel
```
because the former will succeed if err wraps ``gosnmp.ErrUnknownSecurityLevel``.